### PR TITLE
Cleanup and logic fixes

### PR DESCRIPTION
**Pull request by kenshaw**
Logic fixes, code cleanup, and addition of the inverse flag (and implementation).

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -72,97 +55,224 @@ func main() {
...
-	fmt.Fprintf(os.Stdout, "%s%s%s", *flagPrefix, verstr, extra)
+	if *flagInverse == "" {
+		var vers []string
+		if vers, err = getVersion(wd); err != nil {
+			return err
+		}
+		if n := len(vers) - 1; *flagShort && vers[n] == "0" {
+			vers = vers[:n]
+		}
+		fmt.Fprint(os.Stdout, *flagPrefix, strings.Join(vers, *flagSep), extra)
```
Comment (mvdan) [inline main.go:66]: I'd return early here, to keep the rest of the code unindented.

Also, any reason you prefer `fmt.Fprint(os.Stdout, ...)` over `fmt.Print(...)`?

File: main.go [hunk 2]
```diff
@@ -72,97 +55,224 @@ func main() {
...
 		return nil, err
 	}
-	hash, err := repo.ResolveRevision(plumbing.Revision(*flagRev))
+
+	// find commit
+	var hash *plumbing.Hash
+	var commit *object.Commit
+	if hash, err = repo.ResolveRevision(plumbing.Revision(*flagRev)); err != nil {
+		// could not resolve rev, so search for associated object
+		if h := plumbing.NewHash(*flagRev); h != plumbing.ZeroHash {
```
Comment (mvdan) [inline main.go:93]: I'm not clear on why this is necessary. As far as I know, the previous code worked fine. When did it fail incorrectly?

File: main.go [hunk 3]
```diff
@@ -72,97 +55,224 @@ func main() {
...
+	var err error
+
+	// open repository
+	repo, err := git.PlainOpen(wd)
+	if err != nil {
+		return "", err
+	}
+
+	// get head
+	head, err := repo.Head()
```
Comment (mvdan) [inline main.go:154]: Surely this should be using flagRev?

**Review by mvdan — COMMENTED** [review]
This really badly needs tests now, since it's the second time we're refactoring it.

The changes look fine, but left a few comments. Feel free to amend and merge.

File: main.go (hunk 1 above)
Comment (kenshaw) [reply main.go:66]: 1) These don't return exactly similar things -- the primary `get*` func's return different types, for specific reasons. Doing an "early return" would necessitate an additional func, which would needlessly complicate the logic, while not netting any reduction in total lines

2) I would normally use`fmt.Print` in single files where there all output is sent to stdout. However, note that in the main block, the error is sent to stderr. The conscious decision to use `fmt.Fprint` was to make the code more symmetrical throughout. Additionally, `fmt.Print` is just a wrapper for `fmt.Fprint`, and as such you avoid the additional call overhead.

File: main.go (hunk 2 above)
Comment (kenshaw) [reply main.go:93]: You could not pass a specific hash:
```sh
# from the previous version
$ go build && ./git-buildnumber -rev=765065a1317e96bdd6593bf6fa0e2f509777ffd9
reference not found

# from wip-cleanup-fixes branch
$ git checkout wip-cleanup-fixes
$ go build && ./git-buildnumber -rev=765065a1317e96bdd6593bf6fa0e2f509777ffd9
v1.3.13.1
```

File: main.go (hunk 3 above)
Comment (kenshaw) [reply main.go:154]: No -- the inverse would look at what the current head is, and walk backward. `git-buildnumber` wouldn't know what the `-rev` value would be, if you're inversing the version string.

**Review by mvdan — APPROVED** [review]
Please squash/rebase before merging

**Comment by kenshaw** [comment]
Rebased.

**Pull request merged**